gnucash/gnucash-core-app.cpp

```diff
@@ -219,6 +219,11 @@ Gnucash::CoreApp::parse_command_line (int argc, char **argv)
 {
     try
     {
+    // don't process args on and after "--"
+    std::vector<std::string_view> argv_vec (argv, argv + argc);
+    auto match = std::find(argv_vec.begin(), argv_vec.end(), "--");
+    argc = std::distance (argv_vec.begin(), match);
```

Comment:
It's nice to use standard algorithms (as I suggested in my previous review) but the bigger issue with this is that we (well, I...) chose boost::program_options for our command line handling and that this new code is working around it rather than using it to also handle the options to pass to scripts.

I am wary of using two code paths to parse options. Sooner or later they will start to interfere.

We may want to dig into more advanced use of boost::program_options to allow it to also be used to parse options to pass on to a script.

Reply:
`boost::program_options`'s [positional options](https://www.boost.org/doc/libs/1_83_0/doc/html/program_options/overview.html#id-1.3.30.5.7.10) facility gets us partway there and is nicer than the `std::find` hack. You're already using it for the input file and the `std::find` hack might break that for your `--Something.gnucash` example.

Reply:
Ok removed `--` use and use `--script-args abc def ghi` instead. The disadvantage is guile/python won't receive any args which begin with `--` because boost will munge them.

```
chris@pop-os:~/sources/gnucash [env]$ ./stable-install/bin/gnucash-cli --interactive --script-args 1 2 3
This is a development version. It may or may not work.
Report bugs and other problems to gnucash-devel@gnucash.org
You can also lookup and file bug reports at https://bugs.gnucash.org
To find the last stable version, please refer to https://www.gnucash.org/
Welcome to Gnucash Interactive Guile Session
(use-modules (gnucash core-utils))
(use-modules (gnucash engine))
(use-modules (gnucash app-utils))
(use-modules (gnucash report))
(use-modules (system repl repl))
(use-modules (ice-9 readline))
GNU Guile 3.0.9
Copyright (C) 1995-2023 Free Software Foundation, Inc.

Guile comes with ABSOLUTELY NO WARRANTY; for details type `,show w'.
This program is free software, and you are welcome to redistribute it
under certain conditions; type `,show c' for details.

Enter `,help' for help.
scheme@(guile-user)> (program-arguments )
$1 = ("gnucash-cli" "1" "2" "3")
scheme@(guile-user)> 

chris@pop-os:~/sources/gnucash [env]$ ./stable-install/bin/gnucash-cli --interactive --script-args 1 2 3 -L python
This is a development version. It may or may not work.
Report bugs and other problems to gnucash-devel@gnucash.org
You can also lookup and file bug reports at https://bugs.gnucash.org
To find the last stable version, please refer to https://www.gnucash.org/
Welcome to Gnucash Interactive Python Session
>>> import sys
>>> sys.argv
['1', '2', '3']
>>> 
```

Reply:
> The disadvantage is guile/python won't receive any args which begin with -- because boost will munge them.

It doesn't munge them, it tries to parse them, fails, and puts up the usage message unless you either quote them as args to `--script args` or by passing `--` to read them as positional arguments.
```
 bin/gnucash-cli --interactive --script-args "--1 --2 --3"
…
Enter `,help' for help.
scheme@(guile-user)> (program-arguments )
$1 = ("gnucash-cli" "--1 --2 --3")
scheme@(guile-user)>
```
Note that that while that protects the arguments from being parsed by boost::program_options it also turns them into a single option that needs to be separated.
```
bin/gnucash-cli --interactive -- --1 --2 --3
…
Enter `,help' for help.
scheme@(guile-user)> (program-arguments)
$1 = ("gnucash-cli" "--1" "--2" "--3")
scheme@(guile-user)>
```

But now the book file gets included in the script arguments instead of being passed to gnucash-cli to open a session:
```
bin/gnucash-cli myfile.gnucash --interactive --script-args "1 2 3"
…
Enter `,help' for help.
scheme@(guile-user)> (program-arguments)
$1 = ("gnucash-cli" "myfile.gnucash" "1 2 3")
scheme@(guile-user)>
```
And you're not giving the interpreter a shot at the options either:
```
bin/gnucash-cli --interactive -- '-c (program-arguments)' 1 3 3
…
Enter `,help' for help.
scheme@(guile-user)> (program-arguments)
$1 = ("gnucash-cli" "-c (program-arguments)" "1" "3" "3")
scheme@(guile-user)>
```

Reply:
> It doesn't munge them, it tries to parse them, fails, and puts up the usage message unless you either quote them as args to `--script args` or by passing `--` to read them as positional arguments.

> But now the book file gets included in the script arguments instead of being passed to gnucash-cli to open a session:
> 
> ```
> bin/gnucash-cli myfile.gnucash --interactive --script-args "1 2 3"
> …
> Enter `,help' for help.
> scheme@(guile-user)> (program-arguments)
> $1 = ("gnucash-cli" "myfile.gnucash" "1 2 3")
> scheme@(guile-user)>
> ```

I think this is fixed by limiting --input-file to 1 arg: `m_pos_opt_desc.add("input-file", 1);`

> And you're not giving the interpreter a shot at the options either:
> 
> ```
> bin/gnucash-cli --interactive -- '-c (program-arguments)' 1 3 3
> …
> Enter `,help' for help.
> scheme@(guile-user)> (program-arguments)
> $1 = ("gnucash-cli" "-c (program-arguments)" "1" "3" "3")
> scheme@(guile-user)>
> ```

I'm not sure this is possible.

Reply:
Another observation with `--` -- if there's no `--input-file` argument, the first argument after `--` will be stored as the input-file. `gnucash-cli --script script.scm -- datafile.gnucash abc def ghi` will be interpreted as:

```
Script args: abc def ghi
File to load: datafile.gnucash
Language: guile
Script: script.scm
Readwrite: N
Interactive: N
```

So, if the user definitely wants to experiment with `--` with the REPL or script on an empty session, they'll be obliged to input something like `gnucash-cli --script script.scm --interactive -- /dev/null abc def ghi`. I think this is very very unlikely to be a necessity (but ought to be documented somewhere).

Reply:
GnuCash has a --no-file option. If this is set and boost::program_options has set an input_file anyway, you know it really is part of the command line options to be passed to a script. So if the user doesn't need or want to set a file on the command line but wants to pass options to a script, this combination would be the way forward.

This kind of option pre-processing is what I intended to be done in `GnuCash::GnuCashCli::start`.

Reply:
Ok I admit I'm finding this to be tedious... `gnucash` would open the last file by default, and this behaviour is modified via `--no-file`. `gnucash-cli` has no particular business opening the last file by default, and may receive a `--no-file` just to satisfy boost's positional arguments. Would it be sensible for `gnucash-cli` to change its behaviour to open the last file, so that `--no-file` becomes sensible?

Reply:
When I asked earlier on what the added value was of offering full scripting access via gnucash-cli, these kind of complexities were what I was worried about.

As I suggested then I can see the usefulness of opening an interactive shell to quickly run some experiments. I still have my reservations on expanding this to a full blown scripting environment.

The only thing you save is the need to set up the environment. Once that's done you can start a script to load the proper guile/python gnucash modules and have your script interact with the gnucash api the same way as you're now trying to bolt onto gnucash-cli.

Your earlier comment makes it seem like a huge burden to bootstrap gnucash bindings. I have my doubts about that. What's needed to set up our environment is already encapsulated in our environment file. Translate that into a python or guile script that sets a few variables and you're done. Document it in the wiki and it's available to other interested hackers (I recently updated the [python dcoumentation](https://wiki.gnucash.org/wiki/Python_Bindings) btw). Make guile/python equivalents of the environment file and add them as some kind of bootstrap scripts that get installed while installing gnucash (or the relevant bindings) and it's ready to use out of the box.

Personally I perceive gnucash-cli rather as a bottleneck for scripting (though possibly helpful for interactive exploration) rather than a convenience. But that's only my sentiment.

As you are the one willing to spend the effort, I try to limit myself to giving feedback to help you work out a reasonably consistent interface.

So as a direct answer to your question, IMO it's safer if gnucash-cli doesn't automatically load the last file -- particularly when handling scripts. When a file gets loaded automatically in gnucash, you have visual feedback before you take any action. You can see whether the opened file is the one you expect. With scripts, these may automatically make changes without this user interactive verification. By at least requiring an explicit file name (either as gnucash-cli option or written somewhere in the script) you have some control over what gets changed and you can be confident that in subsequent calls of the same command the same file will be opened.

If `--no-file` doesn't feel right, you could also code it such that 
* for interactive mode you expect a file name. One could debate whether or not to autoload the last opened file in the absence of a file name.
* for script mode the script is responsible for loading a file. That would make it more in line with pure scripting. If you really want to pass a file name via the command line, let the script read this file name rather than gnucash.

Reply:
> > And you're not giving the interpreter a shot at the options either:
> > ```
> > bin/gnucash-cli --interactive -- '-c (program-arguments)' 1 3 3
> > …
> > Enter `,help' for help.
> > scheme@(guile-user)> (program-arguments)
> > $1 = ("gnucash-cli" "-c (program-arguments)" "1" "3" "3")
> > scheme@(guile-user)>
> > ```
> 
> I'm not sure this is possible.

I don't think that would be worth the effort. It boils down to reimplementing bash' command line option parser, which converts tokens in a space separated string to an array of tokens. But I's full of special cases (like quoting), and escape sequences. So my suggestion of using a separate option to specify script related command line options is probably not a very good one.

Reply:
Ok -- to keep things very simple and safe, how about we completely forego scripts and REPL getting access to cli arguments?

Reply:
How would you feel as novice hacker wanting to script gnucash, only to find out there are two interfaces to do so that have different limitations ? And example scripts that may work with one interface but not the other with no obvious explanation why ?

Reply:
I think right now there's not much scripting taking place, and we could wait for demand to ramp up to revisit this issue.

From a couple comments above there's also the possibility of combining `--script` to load the script and `--interactive` to load the repl. So it's awkward to determine whether to require a datafile.

Reply:
> Your earlier comment makes it seem like a huge burden to bootstrap gnucash bindings. I have my doubts about that. What's needed to set up our environment is already encapsulated in our environment file. Translate that into a python or guile script that sets a few variables and you're done. Document it in the wiki and it's available to other interested hackers (I recently updated the [python dcoumentation](https://wiki.gnucash.org/wiki/Python_Bindings) btw). Make guile/python equivalents of the environment file and add them as some kind of bootstrap scripts that get installed while installing gnucash (or the relevant bindings) and it's ready to use out of the box.

I've checked the python documentation and I still find it rather confusing to understand where to paste the required paths -- plus, with the myriad of packages or local install, and in reproducible builds, the package dirs to include into python/guile can change (eg /nix/store/long-hex-content-addressable-string/gnucash/bin) . The one constant in many packages that I've seen is they provide easy access to the executable; i.e. we can always rely on them offering access to `gnucash` or `gnucash-cli`; hence the desire for easy access to scripting and run quick experiments on a local datafile.

> So as a direct answer to your question, IMO it's safer if gnucash-cli doesn't automatically load the last file -- particularly when handling scripts. When a file gets loaded automatically in gnucash, you have visual feedback before you take any action. You can see whether the opened file is the one you expect. With scripts, these may automatically make changes without this user interactive verification. By at least requiring an explicit file name (either as gnucash-cli option or written somewhere in the script) you have some control over what gets changed and you can be confident that in subsequent calls of the same command the same file will be opened.

To be mindful of a script damaging a datafile, I've ensured that the datafile is open readonly by default, and only allow read/write if specifically requested on the cli via `--readwrite`. So the following cli will all open the file safely.

```
gnucash-cli datafile.gnucash --interactive
gnucash-cli datafile.gnucash --interactive --language python
gnucash-cli datafile.gnucash --script show-the-balance-sheet.scm
gnucash-cli datafile.gnucash --script my-report.scm --interactive
```

It may be an easier option to remove the datafile option for scripting, and they will be required to load the datafile manually. But I'm now considering how to ensure the scripts can work identically in both gnucash-cli and guile with `GUILE_LOAD_PATH` scenarios.

Reply:
>>> And you're not giving the interpreter a shot at the options either:

>> I'm not sure this is possible.

>  I don't think that would be worth the effort. It boils down to reimplementing bash' command line option parser, which converts tokens in a space separated string to an array of tokens.

It may not be worth the effort, but it has nothing to do with bash or any other shell's option parser. Those are for parsing options when *starting an instance of the shell* and are not invoked when running a program. Shells are REPLs with a command parser just like any other interpreter, including guile and python.  That said, `run_guile_cli`


>> I'm not sure this is possible.

It depends heavily on how you start guile and python3. It's not possible the way you've done it with guile&mdash;you'd need to call `scm_shell` instead of `scm_boot_guile`&mdashbut it may already work with python depending on how the PreInit stuff works. The docs aren't especially clear about that.

Reply:
> > If --no-file doesn't feel right, you could also code it such that
> > 
> > * for interactive mode you expect a file name. One could debate whether or not to autoload the last opened file in the absence of a file name.
> > * for script mode the script is responsible for loading a file. That would make it more in line with pure scripting. If you really want to pass a file name via the command line, let the script read this file name rather than gnucash.
> 
> I also don't like adding `--no-file`. Better to go the other way and provide a `--file` option, but I think it reasonable to try loading the first positional arg and if you get a "no backend" or "file not found" error then assume the user wanted an empty session and the first arg is for the script. You could still provide `--file` for the user to unambiguously specify a datafile, in which case you'd quit with an error on "no backend" or "file not found".

Hmm here's an idea: there's an `--input-file` already for the datafile; how about we attempt to load the datafile if it exists, but also _always_ push the datafile into the scripting engine? i.e.

```cpp
  if (boost::filesystem::is_regular_file (*m_file_to_load))
    load_file (*m_file_to_load);
  // etc
  std::vector newArgv { "gnucash-cli", *m_file_to_load }
```

Reply:
Probably a good idea to pass the datafile name or something to indicate its absense, but [using `operator*()` or `operator->()` on an invalid boost::optional is UB](https://www.boost.org/doc/libs/1_82_0/libs/utility/OptionalPointee.html). `load_file` can just take the boost::optional instead and decide what to do, but you need to put something that the script can understand, maybe `'\0'`, into `newArgv`.

Reply:
Ok. I think testing the datafile argument via boost is_regular_file in inadequate because the data may be in an SQL server.

Reply:
Another radical idea: instead of trying to separate the boost vs scripting args, we could  [accept all arguments](https://www.boost.org/doc/libs/1_82_0/doc/html/program_options/howto.html#id-1.3.30.6.9) and send all args to scripts.

Reply:
> I think testing the datafile argument via boost is_regular_file in inadequate because the data may be in an SQL server.

The current iteration is OK however the `--input-file` argument expects a regular file. I cannot find a convenient `is_valid_input_file` function anywhere which can recognise /path/to/datafile.gnucash and //network/path/datafile.gnucash and mysql://server/datafile

Reply:
> Another radical idea: instead of trying to separate the boost vs scripting args, we could [accept all arguments](https://www.boost.org/doc/libs/1_82_0/doc/html/program_options/howto.html#id-1.3.30.6.9) and send all args to scripts.

The link is for doing the exact opposite: It explains how to separate out the unknown options and pass them on to the script.  That would be an alternative or supplement to the `--` behavior we discussed a few days ago and is a good idea. Passing gnucash-cli's options on to the script and leaving it to the script to figure out would be an unnecessary burden on the script author.

Reply:
> The current iteration is OK however the --input-file argument expects a regular file. I cannot find a convenient is_valid_input_file function anywhere which can recognise /path/to/datafile.gnucash and //network/path/datafile.gnucash and mysql://server/datafile

Good point. If there isn't one already you should be able to piece something together with the functions in libgnucash/core-utils/gnc-path.h and libgnucash/engine/gnc-uri-utils.h.

Reply:
> > Another radical idea: instead of trying to separate the boost vs scripting args, we could [accept all arguments](https://www.boost.org/doc/libs/1_82_0/doc/html/program_options/howto.html#id-1.3.30.6.9) and send all args to scripts.
> 
> The link is for doing the exact opposite: It explains how to separate out the unknown options and pass them on to the script. That would be an alternative or supplement to the `--` behavior we discussed a few days ago and is a good idea.

Or, counterpoint... if the script.scm or .py expects something like `--input-qif file.qif` and gnucash-cli later gains some qif-import powers, then the `--input-qif` will no longer be offered to scripting...?

Reply:
In that case the user would have to change their script to rename their option unless you can figure out a way to limit the options that boost considers depending on other options. IOW if `--interactive` or `--script` are set then `--input-qif` isn't a "known option" and it gets passed on to the script. That's worth working on, I think it's possible now to pass a bunch of options that should be mutually exclusive and which ones get handled is dependent on what order boost::program_options decides to process them.

Reply:
I attemped the boost https://www.boost.org/doc/libs/1_83_0/doc/html/boost/program_options/collect_unrecognized.html and I think it's not the right approach; this will collect only the unrecognized `--optionname` type args and would still require supporting positional args for the unhyphenated options.

Reply:
I agree with the conclusion but for a different reason. Of course it will still be necessary to handle positional options. That's not really a problem and it would free the user from having to separate the `gnucash-cli` options from the script's options with `--`.  But it means that if the user has a typo in a `gnucash-cli` option that it will be passed on to the script instead of displaying the usage message and we'd need to handle unrecognized options for all of the other `gnucash-cli` actions. The complexity of getting that right makes the `--` separator look pretty reasonable. That should be explained in documentation.

Reply:
Last commit restores the `--` behaviour.

Reply:
However I now agree `--` isn't the best approach; currently following are equivalent, and this will be broken when special-casing `--`.

```
gnucash-cli datafile.gnucash --script s.scm
gnucash-cli --input-file datafile.gnucash --script s.scm
gnucash-cli --script s.scm -- datafile.gnucash
```

So I would now suggest using `--script-args` which takes all positional arguments after the input-file.

Reply:
> > The current iteration is OK however the --input-file argument expects a regular file. I cannot find a convenient is_valid_input_file function anywhere which can recognise /path/to/datafile.gnucash and //network/path/datafile.gnucash and mysql://server/datafile
> 
> Good point. If there isn't one already you should be able to piece something together with the functions in libgnucash/core-utils/gnc-path.h and libgnucash/engine/gnc-uri-utils.h.

It's rather difficult because the URI validation logic is mixed up in https://github.com/Gnucash/gnucash/blob/ad15b2db5dd772ff45748f51f2d6310cca47a0c1/libgnucash/engine/qofsession.cpp#L250 onwards.

Reply:
> It's rather difficult because the URI validation logic is mixed up in 

That's the final check and you can rely on that if you want, but I thought you were looking for a "does this look like a file or a URI that might be a book?". The second part is simply screening on the URI scheme being one of `file:`, `mysql:` or `postgres:`; the first part you already had with a boost::filesystem call, don't remember which and you've since removed it. That's the best you can do without trying to start a session. The value is that if it looks like a path/URI but the session won't start you can report it as an error, but if it doesn't then it's reasonable to assume that it should be passed to the script.

Reply:
Ok written to accept these 3 schemes. IMHO if the filename/uri is *always* pushed into argv[1] then we can be very consistent.

Reply:
> Ok written to accept these 3 schemes. IMHO if the filename/uri is _always_ pushed into argv[1] (even the empty filename `""`) then we can be very consistent.

Do you mean as argv[1] to gnucash-cli ? That would make for a command line interface that's not very flexible and not really in the spirit of command line arguments.

This whole command line interface seems to be the biggest challenge of this PR btw... I think the real issue is we want to interpret the command line options and parameters very differently depending on the mode we activate gnucash-cli in.

Current stable has two such modes: a Quotes mode and one Reports mode. This PR adds a third (and possibly fourth) mode, namely Scripting.

The command line parsing as it's implemented so far really is designed to deal only with very simple subcommands in each mode. Each mode switch takes one single parameter subcommand and optionally a file name.

The Scripting mode moves away from that. It can take an arbitrary number of additional command line options and parameters.

Efforts so far suggest these two paradigms are hard to combine.

So here is another option to explore: what if we make each mode more self-contained and make it so each mode gets its own option parser ? I mean perhaps we can create a more hierarchical option parsing scheme where each node in the hierarchy has its own boost::program::options object, only specialized in its own set of options.

Here is how it would work:

There is a top-level option parser, say `mode_parser`. This parser only understands the mode switches (`-Q`, `-R` and whatever we take for interactive/scripting) and a `--help` switch. Any other command line switch is just captured as an unknown option. If a valid mode is detected, this will invoke a second-tier option parser specialized in only that mode. It will be passed all options the top-level parser couldn't parse.

That would help us separate the much more complex Scripting mode from the other modes and would even allow us to define different option conventions for the Scripting mode and hopefully allow for more freedom in what you want to define for Scripting.

This is superficially similar to how git works. You run git <command> <command-specific-options>.

The help function could also be hierarchical. Running `gnucash-cli --help` would explain the modes. Running `gnucash-cli -Q --help` would explain how the Quotes mode works.

Reply:
> > Ok written to accept these 3 schemes. IMHO if the filename/uri is _always_ pushed into argv[1] (even the empty filename `""`) then we can be very consistent.
> 
> Do you mean as argv[1] to gnucash-cli ? That would make for a command line interface that's not very flexible and not really in the spirit of command line arguments.

No, I mean gnucash-cli's args are "rearranged" so that calling `gnucash-cli --script script.scm --input-file datafile.gnucash --script-args 1 2 3` results in guile's `(program-arguments)` returning `'("gnucash-cli" "datafile.gnucash" "1" "2" "3")`.

Otherwise I like the idea of a top-level command with subcommands.

Reply:
I think @gjanssens is on the right track with his modes suggestion. boost::program_options superficially [appears to support it](https://www.boost.org/doc/libs/1_83_0/doc/html/program_options/howto.html#id-1.3.30.6.6) but `bpo::command_line_parser` and `bpo::store` are plain functions so I think we'll have to resort to using [unknown_options](https://www.boost.org/doc/libs/1_83_0/doc/html/program_options/howto.html#id-1.3.30.6.9) to collect the secondary options and then reparse them with the mode's `bpo::options_description` object.

Reply:
Ok perhaps this particular branch must be paused until the stable branch program_options are updated?

Reply:
Perhaps. Rewiring the existing program options handling needs to be a separate commit but that doesn't rule out it being part of this PR.